### PR TITLE
only show visible html + scrolling tool

### PR DESCRIPTION
**Pull request by vnawrath**

### CODE REVIEW

File: ai_powered_qa/custom_plugins/playwright_plugin.py [hunk 1]
```diff
@@ -167,12 +226,45 @@ async def _assert_that(self, selector: str, action: str, value: str = None):
             return "Not implemented action"
         return f"Action '{action}' was successfully performed: {result_message}"
 
+    @tool
+    def scroll(self, selector: str, direction: str):
+        """
+        Scroll up or down in a selected scroll container
+
+        :param str selector: CSS selector for the scroll container
+        :param str direction: Direction to scroll in. Either 'up' or 'down'
```
Comment (vnawrath) [inline ai_powered_qa/custom_plugins/playwright_plugin.py:235]: On some websites, there can be multiple scrollable regions. We mark them for the agent, and let it select the one it wants to scroll in. Then move the mouse there and turn the mouse wheel.

File: ai_powered_qa/custom_plugins/playwright_plugin.py [hunk 2]
```diff
@@ -216,21 +308,39 @@ def clean_html(html_content):
...
+    html_clean = _clean_attributes(html_clean)
     html_clean = re.sub(r"<div[\s]*>[\s]*</div>", "", html_clean)
     html_clean = re.sub(r"<!--[\s\S]*?-->", "", html_clean)
-    html_clean = html_clean.replace("<div", "<d").replace("</div>", "</d>")
-    soup = BeautifulSoup(html_clean, "html.parser")
-    _unwrap_single_child(soup)
-    _remove_useless_tags(soup)
-    return str(soup)
+    # html_clean = html_clean.replace("<div", "<d").replace("</div>", "</d>")
+    return html_clean
```
Comment (vnawrath) [inline ai_powered_qa/custom_plugins/playwright_plugin.py:320]: Reordered to have access to all elements in soup. As the logic get's more complex, we might want to do it all in soup.

File: ai_powered_qa/custom_plugins/playwright_plugin.py [hunk 3]
```diff
@@ -216,21 +308,39 @@ def clean_html(html_content):
...
+
+
+def _remove_not_visible(soup: BeautifulSoup):
+    to_keep = set()
```
Comment (vnawrath) [inline ai_powered_qa/custom_plugins/playwright_plugin.py:324]: We need to make sure all the parent elements of a visible element are visible.

File: web_ui_next.py [hunk 4]
```diff
@@ -8,34 +8,38 @@
...
 
-@st.cache_resource
-def get_agent_store():
-    return AgentStore(
-        "agents",
-        name_to_plugin_class={
-            "PlaywrightPlugin": PlaywrightPlugin,
-        },
-    )
+agent_store = AgentStore(
```
Comment (vnawrath) [inline web_ui_next.py:14]: The agent store doesn't have any state that needs to be cached. Recreating it everytime allows "warm" reloading of plugins which code has changed.

File: web_ui_next.py [hunk 5]
```diff
@@ -8,34 +8,38 @@
...
 
-agent_store = get_agent_store()
+sidebar = st.sidebar
 
 
-@st.cache_resource
-def get_agent(agent_name):
-    new_agent = agent_store.load_agent(
-        agent_name, default_kwargs={"plugins": {"PlaywrightPlugin": PlaywrightPlugin()}}
+def load_agent():
```
Comment (vnawrath) [inline web_ui_next.py:25]: Agent does need to be cached, but we can do it ourselves in session_state. That way we have more control over the logic (e.g. we can implement a "Reload Agent" button.

File: web_ui_next.py [hunk 6]
```diff
@@ -143,9 +147,8 @@ def on_clear_history():
 
 
 with st.chat_message("user"):
-    st.text_area(
-        "Context message", context_message["content"], height=200, disabled=True
-    )
+    st.write("**Cotext message**")
+    st.write(context_message["content"])
```
Comment (vnawrath) [inline web_ui_next.py:151]: I wanted to keep the context message as markdown, as it makes copying the HTML easier.

**Comment by msvana** [comment]
Tried it again on my favorite test subject. The HTML contains stuff that shouldn't be visible:

image

The part after (c) 2024 is at the bottom of the page. It is not visible. I checked both the browser and the screenshot.

In general I think we should also test that filtering out hidden elements and scrolling works better than much the much simpler approach of splitting the HTML into multiple parts.

**Comment by msvana** [comment]
I also noticed that one change I made in the UI improvements is not there. I replaced the context message element with an undeditable textare with fixed witdh. Was this intentional?

**Comment by vnawrath** [comment]
> I also noticed that one change I made in the UI improvements is not there. I replaced the context message element with an undeditable textare with fixed witdh. Was this intentional?

I explained it in my comments. I wanted to keep the markdown formatting of the context message, since we are using the triple ticks notation for the HTML - it makes it easier to copy. 

I assume the textarea was to cap the height of the context message? We should be able to make the HTML a one-liner, making it even shorter than the textarea, and I'd prefer to have better view of other things in the context message (e.g. list of steps to complete, or other information from outside of the PlaywrightPlugin)

**Pull request merged**